### PR TITLE
[Merged by Bors] - feat(Order): the transfinite iteration of a map

### PR DESCRIPTION
**Pull request by joelriou**
Given `φ : I → I` where `[SupSet I]`, we define the `j`th transfinite iteration of `φ` for any `j : J`, with `J` a well-ordered type: this is `transfiniteIterate φ j : I → I`. If `i₀ : I`, then `transfiniteIterate φ ⊥ i₀ = i₀`; if `j` is a non maximal element, than `transfiniteIterate φ (Order.succ j) i₀ = φ (transfiniteIterate φ j i₀)`; and if `j` is a limit element, `transfiniteIterate φ j i₀` is the supremum of the `transfiniteIterate φ l i₀` for `l < j`.

(Hopefully, this will be used in order to show Grothendieck abelian categories have enough injectives.)

Co-authored-by: vi.hdz.p@gmail.com

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by vihdzp** [comment]
Am I correct in saying this is pretty much [`OrdinalApprox.lfpApprox`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/FixedPointApproximants.html#OrdinalApprox.lfpApprox) but generalized to other complete lattices? If so, then we should probably deal with deduplicating them.

**Comment by joelriou** [comment]
> Am I correct in saying this is pretty much [`OrdinalApprox.lfpApprox`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/FixedPointApproximants.html#OrdinalApprox.lfpApprox) but generalized to other complete lattices? If so, then we should probably deal with deduplicating them.

I would be happy if this was a particular case of another construction, but please note that in the applications `φ : I → I` satisfies `i ≤ φ i`, but there is no reason that `φ : I →o I`.

**Comment by vihdzp** [comment]
Well, more precisely, I think `OrdinalApprox.lfpApprox` and `TransfiniteIteration` are both special cases of the following definition:
```lean4
def transfiniteIterate {I J : Type*}
    [SupSet I] [LinearOrder J] [SuccOrder J] [WellFoundedLT J] (f : I → I) (n : J) : I → I :=
  SuccOrder.limitRecOn n
    (fun _ _ ↦ id) (fun _ _ g ↦ f ∘ g) (fun y _ IH ↦ ⨆ x : Set.Iio y, IH x.1 x.2)
```
There's no actual requirement that `f` be monotonic, but if isn't, then the limit case won't necessarily correspond to the limit.

**Comment by joelriou** [comment]
> Well, more precisely, I think `OrdinalApprox.lfpApprox` and `TransfiniteIteration` are both special cases of the following definition:
> 
> ```lean
> def transfiniteIterate {I J : Type*}
>     [SupSet I] [PartialOrder J] [SuccOrder J] [WellFoundedLT J] (f : I → I) (n : J) : I → I :=
>   SuccOrder.limitRecOn n
>     (fun _ _ ↦ id) (fun _ _ g ↦ f ∘ g) (fun y _ IH ↦ ⨆ x : Set.Iio y, IH x.1 x.2)
> ```
> 
> There's no actual requirement that `f` be monotonic, but if isn't, then the limit case won't necessarily correspond to the limit. I'm not sure what works for your application, but I'd imagine that we'd actually want to use something like `limsup` instead of the supremum instead.

Thanks very much for the shorter definition!

I think I still want to use the supremum. Under the assumption `i ≤ φ i`, I show that the function which sends `j` to the `j`th iteration evaluated at `i₀ : I` is a monotone function.

**Comment by vihdzp** [comment]
The advantage with using the `limsup` (or really any more proper notion of a limit on complete lattices) is that it also gives us the dual notion of a transfinite iterate for functions with `i ≤ f i` without having to declare it separately.

**Comment by joelriou** [comment]
> The advantage with using the `limsup` (or really any more proper notion of a limit on complete lattices) is that it also gives us the dual notion of a transfinite iterate for functions with `i ≤ f i` without having to declare it separately.
> 
> If you want we can discuss this further on Zulip. I think it'd be really nice to link this definition to our existing ordinal API.

I am not sure there is a need to unify with `OrdinalApprox.lfpApprox`, as this definition is used only in one file, involves `Ordinal` rather than a well-ordered type `J` (which is very much what we want to consider in category theory, because `J` has a category structure) and in the file, it seems the definitional properties matter. It seems to me using `limsup` would only make things more complicated, and anyway in `OrdinalApprox.lfpApprox`, `⨆` is used, not `limsup`.

File: Mathlib/Order/TransfiniteIteration.lean [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+variable {I : Type u} [SupSet I] (φ : I → I)
+  {J : Type w} [LinearOrder J] [SuccOrder J] [WellFoundedLT J]
+
+/-- The `j`th-iteration of a function `φ : I → I` when `j : J` belongs to
+a well-ordered type. -/
+noncomputable def transfiniteIterate (j : J) : I → I :=
+  SuccOrder.limitRecOn j
+    (fun _ _ ↦ id) (fun _ _ g ↦ φ ∘ g) (fun y _ h ↦ ⨆ (x : Set.Iio y), h x.1 x.2)
+
+lemma transfiniteIterate_bot [OrderBot J] (i₀ : I) :
```
Comment (jcommelin) [inline Mathlib/Order/TransfiniteIteration.lean:47]:
```suggestion
@[simp]
lemma transfiniteIterate_bot [OrderBot J] (i₀ : I) :
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**